servers/zms/src/main/java/com/yahoo/athenz/zms/DBService.java

```diff
@@ -2139,7 +2139,7 @@ boolean addSolutionTemplate(ObjectStoreConnection con, String domainName, String
                 Role templateRole = updateTemplateRole(role, domainName, roleName, templateParams);
                 firstEntry = auditLogSeparator(auditDetails, firstEntry);
                 auditDetails.append(" \"add-role\": ");
-                if (!processRole(con, originalRole, domainName, roleName, templateRole,
+                if (!processRole(con, originalRole, domainName, ZMSUtils.removeDomainPrefix(templateRole.getName(), domainName, ROLE_PREFIX), templateRole,
```

Comment:
I don't believe the change is sufficient to correctly handle variable substitutions in the name. As part of the process command we pass the original role object that was retreived without taking into account the substitution. So while the first template apply command will work fine because the original role does not exist thus it's a new add operation, if you try to re-apply the same template again, we'll pick up and process a wrong original name thus we'll try to execute an add operation again instead of modify. So we need to take into account the substituted name when we retrieve the original role as well.

Same change should be applied to both policy and service blocks as well.